File: algorithm_introduction/ss07_quicksort/cpp/tail_recursive_quicksort.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

int _partition(int *A, int p, int r) {
  int x = A[r];
  int i = p - 1;
  for (int j = p; j < r; ++j) {
    if (A[j] <= x) {
      swap(A[j], A[++i]);
    }
  }
  swap(A[r], A[i + 1]);
  return i + 1;
}

void tail_recursive_quicksort(int *A, int p, int r) {
  while (p < r) {
    int q = _partition(A, p, r);
    tail_recursive_quicksort(A, p, q - 1);
    p = q + 1;
  }
}

struct Arr {
  int *arr;
  int size;

  string arr_str() {
    string s = "[";
    for (int i = 0; i < size; ++i) {
      s += to_string(arr[i]);
      s += ", ";
    }
    s += "]";
    return s;
  }
};

int main() {
  int A[] = {2, 8, 7, 1, 3, 5, 6, 4};
  int B[] = {2, 8, 7, 1, 3, 5, 6, 4, 9};

  Arr arr_A{A, sizeof(A) / sizeof(A[0])};
  Arr arr_B{B, sizeof(B) / sizeof(B[0])};

  unordered_map<string, Arr *> cases = {{"A", &arr_A}, {"B", &arr_B}};
  for (auto &p : cases) {
    cout << "# " << p.first << "\n";
    cout << "before" << p.second->arr_str() << "\n";
    tail_recursive_quicksort(p.second->arr, 0, p.second->size - 1);
    cout << "after" << p.second->arr_str() << "\n" << endl;
  }
}

File: algorithm_introduction/ss07_quicksort/python/tail_recursive_quicksort.py
from array import array


def partition(A: array, p: int, r: int):
    x = A[r]
    i = p - 1
    for j in range(p, r):
        if A[j] <= x:
            i += 1
            A[i], A[j] = A[j], A[i]
    A[i + 1], A[r] = A[r], A[i + 1]
    return i + 1


def tail_recursive_quicksort(A: array, p: int, r: int):
    while (p < r):
        q = partition(A, p, r)
        tail_recursive_quicksort(A, p, q - 1)
        p = q + 1


def main():
    cases = {
        'A': array('I', [2, 8, 7, 1, 3, 5, 6, 4]),
        'B': array('I', [2, 8, 7, 1, 3, 5, 6, 4, 9]),
    }

    for k, v in cases.items():
        print(f'# {k}')
        print(f'before: {v}')
        tail_recursive_quicksort(v, 0, len(v) - 1)
        print(f'after: {v}')


if __name__ == '__main__':
    main()
